mantidproject/mantid
# Fix incorrect length of y data in HeliumAnalyserEfficiency

### Description of work

#### Summary of work

Alters the algorithm calculation so that we use the mid-point of the wavelength bins rather than the bin boundaries when re-calculating the theoretical efficiency. After some discussion with our scientists, the error calculation has also been amended to take into account the uncertainty in the wavelength bins (see https://github.com/mantidproject/mantid/issues/37941#issuecomment-2360465237 for more details).




Fixes #37941 




#### Further detail of work

In addition to the various calculation changes, this PR also includes the following:

- A bit of a refactor to help with readability in some places and add relevant docstrings to methods. I've also made changes so that we perform the re-calculation of the theoretical efficiency in place, to avoid having to create a new final output workspace.

- Some additional validation to check that all input workspaces contain a single spectrum and are histogram data.

- Documentation updates, including some changes across all of the documentation for the Polarised SANS algorithms to ensure that we're using consistent notation for equations.

- Updates to relevant tests, including system test reference data.


### To test:



Build the `docs-html` target to check the documentation changes.

To test the algorithm result, run this script:

```
wsPara = CreateSampleWorkspace('Histogram', Function='User Defined', UserDefinedFunction='name=UserFunction,Formula=0.5*exp(-0.0733*12*x*(1-0.9))',XUnit='Wavelength', xMin='1',XMax='8', BinWidth='1', NumBanks='1', BankPixelWidth='1')
wsPara1 = CloneWorkspace(wsPara)
wsAnti = CreateSampleWorkspace('Histogram', Function='User Defined', UserDefinedFunction='name=UserFunction,Formula=0.5*exp(-0.0733*12*x*(1+0.9))',XUnit='Wavelength', xMin='1',XMax='8', BinWidth='1', NumBanks='1', BankPixelWidth='1')
wsAnti1 = CloneWorkspace(wsAnti)

grp = GroupWorkspaces([wsPara,wsAnti,wsPara1,wsAnti1])
eff = HeliumAnalyserEfficiency(grp, SpinStates='11,01,00,10')
```

If you view the data in the output workspace (using Show Data), the workspace should have 6 columns under the Y data tab and 7 columns under the X data tab.

Feel free to check the additional workspace validation by attempting to pass in workspaces with more than one spectrum, or workspaces that have been converted to point data (you can use the `ConvertToPointData` algorithm to do this).





---

### Reviewer

Please comment on the points listed below ([full description](http://developer.mantidproject.org/ReviewingAPullRequest.html)).
**Your comments will be used as part of the gatekeeper process, so please comment clearly on what you have checked during your review.** If changes are made to the PR during the review process then your final comment will be the most important for gatekeepers. In this comment you should make it clear why any earlier review is still valid, or confirm that all requested changes have been addressed.

#### Code Review

- Is the code of an acceptable quality?
- Does the code conform to the [coding standards](http://developer.mantidproject.org/Standards/)?
- Are the unit tests small and test the class in isolation?
- If there is GUI work does it follow the [GUI standards](http://developer.mantidproject.org/Standards/GUIStandards.html)?
- If there are changes in the release notes then do they describe the changes appropriately?
- Do the release notes conform to the [release notes guide](https://developer.mantidproject.org/Standards/ReleaseNotesGuide.html)?

#### Functional Tests

- Do changes function as described? Add comments below that describe the tests performed?
- Do the changes handle unexpected situations, e.g. bad input?
- Has the relevant (user and developer) documentation been added/updated?

Does everything look good? Mark the review as **Approve**. A member of `@mantidproject/gatekeepers` will take care of it.

### Gatekeeper

If you need to request changes to a PR then please add a comment and set the review status to "Request changes". This will stop the PR from showing up in the list for other gatekeepers.
